vcd/resource_vcd_vapp_firewall_rules.go

```diff
@@ -0,0 +1,394 @@
+		case "icmp":
+			protocol = &types.FirewallRuleProtocols{
+				ICMP: true,
+			}
+		case "tcp&udp":
+			protocol = &types.FirewallRuleProtocols{
+				TCP: true,
+				UDP: true,
+			}
+		default:
+			protocol = &types.FirewallRuleProtocols{
+				Any: true,
+			}
+		}
+		rule := &types.FirewallRule{
+			IsEnabled:            configuredRule["enabled"].(bool),
+			MatchOnTranslate:     false,
+			Description:          configuredRule["description"].(string),
+			Policy:               configuredRule["policy"].(string),
+			Protocols:            protocol,
+			Port:                 getNumericPort(configuredRule["destination_port"]),
+			DestinationPortRange: strings.ToLower(configuredRule["destination_port"].(string)),
+			DestinationIP:        strings.ToLower(configuredRule["destination_ip"].(string)),
+			SourcePort:           getNumericPort(configuredRule["source_port"]),
+			SourcePortRange:      strings.ToLower(configuredRule["source_port"].(string)),
+			SourceIP:             strings.ToLower(configuredRule["source_ip"].(string)),
+			EnableLogging:        configuredRule["enable_logging"].(bool),
+		}
+
+		if configuredRule["destination_vm_id"].(string) != "" {
+			vm, err := vapp.GetVMById(configuredRule["destination_vm_id"].(string), false)
+			if err != nil {
+				return nil, fmt.Errorf("error fetchining VM: %s", err)
+			}
+
+			rule.DestinationVM = &types.VMSelection{VAppScopedVMID: vm.VM.VAppScopedLocalID,
+				VMNicID: configuredRule["destination_vm_nic_id"].(int), IPType: configuredRule["destination_vm_ip_type"].(string)}
+		}
+		if configuredRule["source_vm_id"].(string) != "" {
+			vm, err := vapp.GetVMById(configuredRule["source_vm_id"].(string), false)
+			if err != nil {
+				return nil, fmt.Errorf("error fetchining VM: %s", err)
+			}
+
+			rule.SourceVM = &types.VMSelection{VAppScopedVMID: vm.VM.VAppScopedLocalID,
+				VMNicID: configuredRule["source_vm_nic_id"].(int), IPType: configuredRule["source_vm_ip_type"].(string)}
+		}
+		firewallRules = append(firewallRules, rule)
+	}
+
+	return firewallRules, nil
+}
+
+// vappFirewallRuleImport is responsible for importing the resource.
+// The following steps happen as part of import
+// 1. The user supplies `terraform import _resource_name_ _the_id_string_` command
+// 2. `_the_id_string_` contains a dot formatted path to resource as in the example below
+// 3. The functions splits the dot-formatted path and tries to lookup the object
+// 4. If the lookup succeeds it set's the ID field for `_resource_name_` resource in state file
+// (the resource must be already defined in .tf config otherwise `terraform import` will complain)
+// 5. `terraform refresh` is being implicitly launched. The Read method looks up all other fields
+// based on the known ID of object.
+//
+// Example resource name (_resource_name_): vcd_vapp_firewall_rules.my_existing_firewall_rules
+// Example import path (_the_id_string_): org.my_existing_vdc.vapp_name.network_name or org.my_existing_vdc.vapp_id.network_id
```

Comment:
I am confused.
This is the same path used to retrieve a vApp network.
What's the functional difference between this function and `resourceVcdVappNetworkImport`?

Reply:
this one handles name or Id. resourceVcdVappNetworkImport only name. Also setts different fields to state file.

Reply:
> This is the same path used to retrieve a vApp network.

@dataclouder The target resource is different though.